### ISSUE TITLE
Gradient computation doesn't work correctly if every sample in the batch uses a different set of weights (gradients are reordered)

### ISSUE THREAD

**Issue by mitkeyastromouse** [keras-team/keras#9406]
Keras version: 2.1.4
Tensorflow version: 1.5.0 (issue can be reproduced with Theano backend as well)
OSX 10.13.3 

I created a custom layer that adds a trainable tensor to the input. The tensor has the same shape as the input batch, including dimension 0. Therefore each sample in the batch gets different variables added to them.

When I run one training step, the gradients are wrong. I mean -- they are correct, but they are randomly shuffled along dimension 0. Therefore they do not correspond to their samples and training therefore doesn't work (the idea is to keep using the same batch on the input during the whole training).

The workaround is of course to train with batch_size=1 but it's not ideal.

Code here:
https://gist.github.com/mitkeyastromouse/5ae860cbc569c12b9125d5903c4a1827

**Comment by myutwo150** [comment]
Turn off shuffle.
```python
model.fit(training_data, training_labels, batch_size=batch_size, epochs=1, shuffle=False)
```

**Comment by myutwo150** [comment]
Turn off shuffle.
```python
model.fit(training_data, training_labels, batch_size=batch_size, epochs=1)
```
